clients/pkg/promtail/targets/syslog/transport.go

```diff
@@ -0,0 +1,375 @@
+				backoff.Wait()
+				continue
+			}
+
+			level.Error(l).Log("msg", "failed to accept syslog connection. quiting", "err", err)
+			return
+		}
+		backoff.Reset()
+
+		t.openConnections.Add(1)
+		go t.handleConnection(c)
+	}
+
+}
+
+func (t *TCPTransport) handleConnection(cn net.Conn) {
+	defer t.openConnections.Done()
+
+	c := &idleTimeoutConn{cn, t.idleTimeout()}
+
+	handlerCtx, cancel := context.WithCancel(t.ctx)
+	defer cancel()
+	go func() {
+		<-handlerCtx.Done()
+		_ = c.Close()
+	}()
+
+	lbs := t.connectionLabels(ipFromConn(c).String())
+
+	err := syslogparser.ParseStream(c, func(result *syslog.Result) {
+		if err := result.Error; err != nil {
+			t.handleMessageError(err)
+			return
+		}
+		t.handleMessage(lbs.Copy(), result.Message)
+	}, t.maxMessageLength())
+
+	if err != nil {
+		level.Warn(t.logger).Log("msg", "error initializing syslog stream", "err", err)
+	}
+}
+
+// Close implements SyslogTransport
+func (t *TCPTransport) Close() error {
+	t.baseTransport.close()
+	return t.listener.Close()
+}
+
+// Wait implements SyslogTransport
+func (t *TCPTransport) Wait() {
+	t.openConnections.Wait()
+}
+
+// Addr implements SyslogTransport
+func (t *TCPTransport) Addr() net.Addr {
+	return t.listener.Addr()
+}
+
+type UDPTransport struct {
+	*baseTransport
+	packetConn net.PacketConn
+	readerPool sync.Pool
+}
+
+func NewSyslogUDPTransport(config *scrapeconfig.SyslogTargetConfig, handleMessage handleMessage, handleError handleMessageError, logger log.Logger) Transport {
+	return &UDPTransport{
+		baseTransport: newBaseTransport(config, handleMessage, handleError, logger),
+		readerPool: sync.Pool{New: func() interface{} {
+			return bytes.NewReader([]byte{})
+		}},
+	}
+}
+
+// Run implements SyslogTransport
+func (t *UDPTransport) Run() error {
+	l, err := net.ListenPacket(protocolUDP, t.config.ListenAddress)
+	if err != nil {
+		return fmt.Errorf("error setting up syslog target: %w", err)
+	}
+	t.packetConn = l
+	level.Info(t.logger).Log("msg", "syslog listening on address", "address", t.Addr().String(), "protocol", protocolUDP)
+	t.openConnections.Add(1)
+	go t.acceptPackets()
+	return nil
+}
+
+// Close implements SyslogTransport
+func (t *UDPTransport) Close() error {
+	t.baseTransport.close()
+	return t.packetConn.Close()
+}
+
+func (t *UDPTransport) acceptPackets() {
+	defer t.openConnections.Done()
+
+	var (
+		n    int
+		addr net.Addr
+		err  error
+	)
+	buf := make([]byte, t.maxMessageLength())
+	for {
+		if !t.Ready() {
+			level.Info(t.logger).Log("msg", "syslog server shutting down")
+			return
+		}
+		n, addr, err = t.packetConn.ReadFrom(buf)
+		if err != nil {
+			level.Info(t.logger).Log("msg", "failed to read packets", "addr", addr, "n", n, "err", err)
+			continue
+		}
+
+		// TODO: is there a better way to pass a copy to the handleRcv() function?
+		b := make([]byte, n)
+		copy(b, buf[:n])
+		r := t.readerPool.Get().(*bytes.Reader)
+		r.Reset(b)
+		t.openConnections.Add(1)
+		go t.handleRcv(addr, r)
```

Comment:
@cyriltovena Is there a better way to do this without doing too many allocations?

I even think it does not make much difference to just pass `b` as `[]byte` to `t.handleRcv()` and create a new reader instance in the goroutine.